### PR TITLE
gh-113246: Updated bundled pip to 23.3.2

### PR DESCRIPTION
**Pull request by sbidoul**
* Issue: gh-113246

### CODE REVIEW

**Comment by hugovk** [comment]
Please could you also update the pip version number here?

https://github.com/python/cpython/blob/f428c4dafbfa2425ea056e7f2ed2ea45fa90be87/Tools/build/generate_sbom.py#L53

And then run `make regen-sbom`.

This is brand new tooling to generate a software bill-of-materials (SBOM), and this PR may be the first test of updating it :)  Tracking issue: https://github.com/python/cpython/issues/112302.

cc @SethMichaelLarson

**Comment by hugovk** [comment]
We'll get this documented: https://github.com/python/devguide/issues/1241

In the meantime, if you're on macOS or Linux and get this:
```
❯ make regen-sbom
make: *** No rule to make target `regen-sbom'.  Stop.
```
Run `./configure` first (https://devguide.python.org/).

If you're on Windows, run `python Tools/build/generate_sbom.py`

**Comment by pfmoore** [comment]
> Please could you also update the pip version number here?

I would suggest updating it to say

` include=["Lib/ensurepip/_bundled/pip-*-py3-none-any.whl"]`

(assuming that works). I *don't* want us to have to manually update this every pip release - it's just another manual task that could trip people up.

File: Tools/build/generate_sbom.py [hunk 1]
```diff
@@ -50,7 +50,7 @@ class PackageFiles(typing.NamedTuple):
         include=["Modules/expat/**"]
     ),
     "pip": PackageFiles(
-        include=["Lib/ensurepip/_bundled/pip-23.3.1-py3-none-any.whl"]
+        include=["Lib/ensurepip/_bundled/pip-23.3.2-py3-none-any.whl"]
```
Comment (pfmoore) [inline Tools/build/generate_sbom.py]:
```suggestion
        include=["Lib/ensurepip/_bundled/pip-*-py3-none-any.whl"]
```

See my comment (you made this change while I was writing it 😉)

Comment (sbidoul) [reply Tools/build/generate_sbom.py]: That seems to work.

**Review by pfmoore — APPROVED** [review]

**Comment by pfmoore** [comment]
@sbidoul I've triggered the backport PRs, but they'll probably need fixing for the SBOM stuff. Is that something you can do?

**Comment by sethmlarson** [comment]
Apologies for putting in a roadblock without the full documentation in place, I've [submitted this PR to the devguide](https://github.com/python/devguide/pull/1251) to remedy that situation.

So this PR unfortunately has left the SBOM in an invalid state, pip has been upgraded to `23.2.2` but the SBOM still lists `versionInfo: 23.2.1`. The change to use `*` would only propagate that invalid state to future upgrades as well since CI no longer fails when an update is made.

Luckily pip (unlike most vendored dependencies) is a part of a packaging ecosystem so is easier to get the version information automatically. I've created [this issue](https://github.com/python/cpython/issues/113257) to work on automating more of this process where possible.

Happy to answer any questions about the SBOM work, this is one of the challenges of trying to add security features while minimizing the impact to core developers work so I'm looking to learn more from you all as well on what would work best.

**Comment by pfmoore** [comment]
Thanks - sorry for merging without your input, as I said I assumed automerge would wait for any requested reviewers. My bad.

Ideally, I'd hope that ultimately adding a new release of pip would mean simply updating the ensurepip `__init__.py` and dropping the new wheel in, as it did in the past. That was what I'd intended by suggesting using the wildcard. Was it the wildcard that made the `versionInfo` field invalid, or is that a different edit that we didn't know to make?

Hopefully, your ultimate resolution will be to automate all of the SBOM generation from just the wheel filename. The format of a wheel filename is standardised, and you can extract the version from it, so I'd hope that would be possible (even if it needed a small amount of special-case code for ensurepip).

**Comment by sethmlarson** [comment]
@pfmoore No worries at all! I also didn't know that was a thing so now I've learned as well.

> Hopefully, your ultimate resolution will be to automate all of the SBOM generation from just the wheel filename. The format of a wheel filename is standardised, and you can extract the version from it, so I'd hope that would be possible (even if it needed a small amount of special-case code for ensurepip).

This is my vision as well, potentially with an added run of `make regen-sbom` after dropping in the wheel so the generation can take place. Would that be acceptable as a workflow?

**Comment by pfmoore** [comment]
> Would that be acceptable as a workflow?

Yep, assuming there's a Windows equivalent of that (i.e. something that doesn't rely on make). I'm bound to forget to do it a few times, and I don't know where would be a good place to document it (thinking in terms of where I look when I need to remember the process) but we can sort details like that out.

**Comment by hugovk** [comment]
> So this PR unfortunately has left the SBOM in an invalid state, pip has been upgraded to `23.2.2` but the SBOM still lists `versionInfo: 23.2.1`.

@sethmlarson Would you like this fixed for tomorrow's 3.13a3 release? If so, maybe fix directly, and update automation later if that will take a little longer.

**Comment by sethmlarson** [comment]
@hugovk I'll submit a PR quickly, thanks for the ping :)

**Comment by sethmlarson** [comment]
Created https://github.com/python/cpython/pull/113262 to fix the pip SBOM metadata.

**Comment by sethmlarson** [comment]
@pfmoore @sbidoul pip SBOM metadata discovery is now automated: https://github.com/python/cpython/pull/113295

**Pull request merged**